// pokemon-showdown/.server-dist/minor-activity.js
"use strict";Object.defineProperty(exports, "__esModule", {value: true});/**
 * Minor activities
 * Pokemon Showdown - http://pokemonshowdown.com/
 *
 * Minor activities are representations of non-game activities that rooms
 * regularly use, such as polls and announcements. Rooms are limited to
 * one minor activity at a time.
 *
 * Minor activities keep track of users in the form of userids and IPs.
 * If a player votes for a poll under one IP, they cannot vote for the same
 * poll again.
 *
 * The user-tracking system is not implemented at the base level: Announcements
 * do not have a reason to keep track of users' IPs/IDs because they're just used
 * to broadcast a message to a room.
 *
 * @license MIT
 */

// globally Rooms.MinorActivity
 class MinorActivity {
	
	
	
	
	
	
	
	constructor(room) {
		this.roomid = room.roomid;
		this.room = room;
		this.name = 'Minor Activity';
		this.timeout = null;
		this.timeoutMins = 0;
		this.supportHTML = false;
	}
	destroy() {}
	save() {}
	



} exports.MinorActivity = MinorActivity;
